llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8789,8 +9029,535 @@ static bool combineCCMask(SDValue &CCReg, int &CCValid, int &CCMask) {
+      OrOp1->getOpcode() == SystemZISD::SELECT_CCMASK) {
+    auto *Op0CCValid = dyn_cast<ConstantSDNode>(OrOp0->getOperand(2));
+    auto *Op0CCMask = dyn_cast<ConstantSDNode>(OrOp0->getOperand(3));
+    auto *Op1CCValid = dyn_cast<ConstantSDNode>(OrOp1->getOperand(2));
+    auto *Op1CCMask = dyn_cast<ConstantSDNode>(OrOp1->getOperand(3));
+    if (!Op0CCValid || !Op1CCValid || !Op0CCMask || !Op1CCMask)
+      return SDValue();
+    int Op0CCValidVal = Op0CCValid->getZExtValue();
+    int Op1CCValidVal = Op1CCValid->getZExtValue();
+    if (Op0CCValidVal != Op1CCValidVal)
+      return SDValue();
+
+    // OrOp0 and OrOp0 should point to same CC.
+    SDValue Op0CCReg = OrOp0->getOperand(4), Op1CCReg = OrOp1->getOperand(4);
+    auto *CCNode0 = Op0CCReg.getNode(), *CCNode1 = Op1CCReg.getNode();
+    if (!CCNode0 || !CCNode1 || CCNode0 != CCNode1)
+      return SDValue();
+
+    // Optimizing only the case where Op0TrueVal and Op1TrueVal are equal
+    // and at the same time Op0FalseVal and Op1FalseVal are also equal.
+    auto *Op0TrueVal = dyn_cast<ConstantSDNode>(OrOp0->getOperand(0));
+    auto *Op0FalseVal = dyn_cast<ConstantSDNode>(OrOp0->getOperand(1));
+    auto *Op1TrueVal = dyn_cast<ConstantSDNode>(OrOp1->getOperand(0));
+    auto *Op1FalseVal = dyn_cast<ConstantSDNode>(OrOp1->getOperand(1));
+    if (!Op0TrueVal || !Op0FalseVal || !Op1TrueVal || !Op1FalseVal)
+      return SDValue();
+    if (Op0TrueVal->getZExtValue() != Op1TrueVal->getZExtValue() ||
+        Op0FalseVal->getZExtValue() != Op1FalseVal->getZExtValue())
+      return SDValue();
+
+    // Compute the effective CC mask for select.
+    int Op0CCMaskVal = Op0CCMask->getZExtValue();
+    int Op1CCMaskVal = Op1CCMask->getZExtValue();
+    // Overlap.
+    if ((Op0CCMaskVal & Op1CCMaskVal) ||
+        !(Op0CCMaskVal & (Op1CCMaskVal ^ SystemZ::CCMASK_ANY)) ||
+        !((Op0CCMaskVal ^ SystemZ::CCMASK_ANY) & Op1CCMaskVal) ||
+        !((Op0CCMaskVal ^ SystemZ::CCMASK_ANY) &
+          (Op1CCMaskVal ^ SystemZ::CCMASK_ANY)))
+      return SDValue();
+    int CCMask = Op0CCMaskVal | Op1CCMaskVal;
+
+    // Check Or's user's user if it it select_ccmask or br_ccmask, put it back
+    // to WorkList so that algortirm can start processing from outer
+    // select_ccmask/br_ccmask and combine with (icmp (select_ccmask)).
+    for (SDUse &IcmpUse : N->uses()) {
+      auto *Icmp = IcmpUse.getUser();
+      if (Icmp && Icmp->getOpcode() == SystemZISD::ICMP) {
+        for (SDUse &SelectBrUse : Icmp->uses()) {
+          auto *SelectBr = SelectBrUse.getUser();
+          if (SelectBr && (SelectBr->getOpcode() == SystemZISD::SELECT_CCMASK ||
+                           SelectBr->getOpcode() == SystemZISD::BR_CCMASK))
+            DCI.AddToWorklist(SelectBr);
+        }
+      }
+    }
+    return DAG.getNode(SystemZISD::SELECT_CCMASK, SDLoc(N), N->getValueType(0),
+                       OrOp0->getOperand(0), OrOp0->getOperand(1),
```

Comment:
Given the mask selection above, here it is the true case value that is not correct: it should be `(or TRUEVAL FALSEVAL)` instead.